Refactor record_move

HotMercury opened the pull request:
Replace static array declaration with dynamic allocation for move_record

In this pull request, I propose replacing the static declaration of the move_record array with dynamic allocation. The static declaration used to allocate space for a fixed number of moves, assuming a 4x4 game board with 16 possible moves. However, not every game may utilize all 16 moves, leading to potential wasted memory. By dynamically allocating memory based on the actual number of moves, we can optimize space usage and enhance the program's efficiency.

HotMercury committed:
Refactor record_move

Dynamically allocate move_record array to conserve space.

jserv commented on ttt.c:
```diff
@@ -15,12 +15,19 @@ static int no_step[2] = {0, 0};
+
+    if (move_record == NULL)
+        exit(1);
+    else {
```
Remove this `else` statement since it is not necessary.

jserv commented on ttt.c:
```diff
@@ -15,12 +15,19 @@ static int no_step[2] = {0, 0};
+    if (move_count == 0)
+        move_record = malloc(sizeof(int));
+    else
+        move_record = realloc(move_record, sizeof(int) * (move_count + 1));
```
Frequent invocation of `realloc` may impact the performance. You can present certain budget for recorded movements in advance.

HotMercury committed:
Reducing the frequency of realloc invocations

The initial size of `move_record` is set to 5 because
a minimum of 5 moves is required to determine the winner.
Subsequent resizing `move_record` is base on the board size,
so it can reduce the frequency of `realloc`.

jserv commented on ttt.c:
```diff
@@ -20,16 +20,16 @@ static int move_count = 0;
-    else
-        move_record = realloc(move_record, sizeof(int) * (move_count + 1));
+    if (move_count == 0) {
+        move_record = malloc(sizeof(int) * (5));
```
Why 5?

HotMercury replied:
The initial size of `move_record` is set to 5 because a minimum of 5 moves is required to determine the winner.

And then in a 4 x 4 board, if both players aim to win, the game can always conclude within 5 moves. Therefore, it is currently unclear what the optimal size for resizing.

jserv replied:
Add the consideration in the comments.

HotMercury replied:
I have add my consideration process into code comments and utilized TODO comments to highlight areas for potential improvement.

jserv commented on ttt.c:
```diff
@@ -20,16 +20,16 @@ static int move_count = 0;
+    }
 
-    if (move_record == NULL)
+    if (move_record == NULL) {
```
Simply rewrite as following:
```c
if (!move_record)
    exit(1);
```

HotMercury committed:
Fixing style

HotMercury committed:
Add comment in record_move

The initial size of move_record is set to 5 because a minimum
of 5 moves is required to determine the winner.

And then in a 4 x 4 board, if both players aim to win, the
game can always conclude within 5 moves. Therefore,
it is currently unclear what the optimal size for resizing.

Todo : find the optimal size.

jserv commented:
Thank @HotMercury for contributing!

idoleat commented:
I have some thoughts on recording gameplay: for wider usage, recording game states could be an option. Here are several reasons
* Gameplay can be replayed from any states, backward and forward. By using move sequence to reach certain state in the gameplay, moves need to be aggregated from start. Viewing gameplays in Counter Strike: Global Offensive suffers from long delay when jumping between rounds, which is caused by replaying from start everytime on jumping. In our case, the gameplay of m, n, k game could be long if k is large enough.
* In case errors can happen while recording, faults only remain in some states rather than whole gameplay, comparing to using moves.
* A disadvantages is that it may take more spaces to store, comparing to 4 bytes at most per move.

Some ideas on designing encoding: Since later states contain earlier states, we may have the code partial ordered. This may make states search-able. Puffer code for encoding a tree is designed to proved the count of all possible trees. This can be a reference. Or just compress the raw array data by compression algorithms like LZ4, which kernel already provides. Perfect hash functions may be considered if the count of possible game states is not too large.

jserv commented:
> I have some thoughts on recording gameplay: for wider usage, recording game states could be an option. Here are several reasons

Create a new issue for tracking.
